File: packages/js/node/src/swap/fixedfloat-rates.ts
/**
 * FixedFloat public XML rates export — the bulk feed for all pairs.
 *
 * Docs: GET https://ff.io/rates/fixed.xml (and float.xml). No API key, no
 * weight budget. OpenReceive caches the parsed index in openreceive_meta and
 * derives indicative quotes / min-max locally. `/create` remains authoritative.
 */

import type { SwapRateType } from "./rates-cache.ts";

export interface FixedFloatRatePair {
  readonly from: string;
  readonly to: string;
  /** Reference send amount in `from` (defines the rate with `out`). */
  readonly in: string;
  /** Reference receive amount in `to`. */
  readonly out: string;
  /** Reserve availability hint for `to` — not a max exchange amount. */
  readonly amount: string;
  readonly minamount: string;
  readonly maxamount: string;
  /** Network fee on the `to` side, excluded from `out` (e.g. "0.0005 BTC"). */
  readonly tofee?: string;
}

export interface FixedFloatRatesIndex {
  readonly fetched_at: number;
  /** Keyed as `${from.toUpperCase()}:${to.toUpperCase()}`. */
  readonly pairs: Readonly<Record<string, FixedFloatRatePair>>;
}

const DECIMAL_PATTERN = /^[0-9]+(\.[0-9]+)?$/;
const SATS_PER_BTC = 100_000_000n;

export function fixedFloatRatesPairKey(from: string, to: string): string {
  return `${from.trim().toUpperCase()}:${to.trim().toUpperCase()}`;
}

export function fixedFloatRatesXmlPath(rateType: SwapRateType = "fixed"): string {
  return `/rates/${rateType}.xml`;
}

export async function fetchFixedFloatRatesIndex(input: {
  readonly baseUrl: string;
  readonly rateType?: SwapRateType;
  readonly fetch: typeof globalThis.fetch;
  readonly now: () => number;
  readonly requestTimeoutMs?: number;
}): Promise<FixedFloatRatesIndex> {
  const rateType = input.rateType ?? "fixed";
  const url = `${input.baseUrl.replace(/\/+$/, "")}${fixedFloatRatesXmlPath(rateType)}`;
  const timeoutMs = input.requestTimeoutMs ?? 10_000;
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), timeoutMs);
  let response: Response;
  try {
    response = await input.fetch(url, {
      method: "GET",
      redirect: "follow",
      signal: controller.signal,
      headers: { Accept: "application/xml, text/xml, */*" },
    });
  } catch (error) {
    const aborted =
      error instanceof Error &&
      (error.name === "AbortError" || error.message.toLowerCase().includes("abort"));
    throw new Error(
      aborted
        ? `FixedFloat rates ${rateType}.xml request timed out.`
        : `FixedFloat rates ${rateType}.xml request failed before a response was received.`,
      { cause: error },
    );
  } finally {
    clearTimeout(timeout);
  }
  if (!response.ok) {
    throw new Error(`FixedFloat rates ${rateType}.xml failed with HTTP ${response.status}.`);
  }
  const xml = await response.text();
  return {
    fetched_at: input.now(),
    pairs: parseFixedFloatRatesXml(xml),
  };
}

export function parseFixedFloatRatesXml(xml: string): Readonly<Record<string, FixedFloatRatePair>> {
  const pairs: Record<string, FixedFloatRatePair> = {};
  for (const itemXml of matchTags(xml, "item")) {
    const from = readTagText(itemXml, "from");
    const to = readTagText(itemXml, "to");
    const inAmount = readTagText(itemXml, "in");
    const outAmount = readTagText(itemXml, "out");
    const amount = readTagText(itemXml, "amount");
    const minamount = readTagText(itemXml, "minamount");
    const maxamount = readTagText(itemXml, "maxamount");
    if (
      from === undefined ||
      to === undefined ||
      inAmount === undefined ||
      outAmount === undefined ||
      amount === undefined ||
      minamount === undefined ||
      maxamount === undefined
    ) {
      continue;
    }
    const tofee = readTagText(itemXml, "tofee");
    const pair: FixedFloatRatePair = {
      from: from.trim(),
      to: to.trim(),
      in: stripCurrencySuffix(inAmount),
      out: stripCurrencySuffix(outAmount),
      amount: stripCurrencySuffix(amount),
      minamount: stripCurrencySuffix(minamount),
      maxamount: stripCurrencySuffix(maxamount),
      ...(tofee === undefined ? {} : { tofee: tofee.trim() }),
    };
    pairs[fixedFloatRatesPairKey(pair.from, pair.to)] = pair;
  }
  return pairs;
}

export function serializeFixedFloatRatesIndex(index: FixedFloatRatesIndex): string {
  return JSON.stringify({
    fetched_at: index.fetched_at,
    pairs: index.pairs,
  });
}

export function deserializeFixedFloatRatesIndex(value: string): FixedFloatRatesIndex {
  const parsed = JSON.parse(value) as {
    readonly fetched_at?: unknown;
    readonly pairs?: unknown;
  };
  if (
    typeof parsed.fetched_at !== "number" ||
    !Number.isSafeInteger(parsed.fetched_at) ||
    parsed.pairs === null ||
    typeof parsed.pairs !== "object" ||
    Array.isArray(parsed.pairs)
  ) {
    throw new Error("Invalid FixedFloat rates cache blob.");
  }
  const pairs: Record<string, FixedFloatRatePair> = {};
  for (const [key, raw] of Object.entries(parsed.pairs as Record<string, unknown>)) {
    const pair = readStoredPair(raw);
    if (pair !== undefined) pairs[key] = pair;
  }
  return { fetched_at: parsed.fetched_at, pairs };
}

/**
 * Indicative pay-in amount for a Lightning payout of `invoiceAmountMsats`,
 * using the XML reference rate (`in`/`out`) and optional BTC `tofee`.
 *
 * Formula (direction=to): pay_from = (invoice_btc + tofee_btc) × (in / out).
 * Rounds the pay-in amount up at 8 decimal places so the UI never understates
 * what `/create` is likely to require.
 */
export function quotePayAmountFromFixedFloatRate(input: {
  readonly pair: FixedFloatRatePair;
  readonly invoiceAmountMsats: number;
}): string | undefined {
  if (!Number.isSafeInteger(input.invoiceAmountMsats) || input.invoiceAmountMsats <= 0) {
    return undefined;
  }
  const rateIn = parsePositiveDecimal(input.pair.in);
  const rateOut = parsePositiveDecimal(input.pair.out);
  if (rateIn === undefined || rateOut === undefined) return undefined;

  const invoiceSats = BigInt(Math.ceil(input.invoiceAmountMsats / 1000));
  const tofeeSats = parseToFeeBtcSats(input.pair.tofee) ?? 0n;
  const totalSats = invoiceSats + tofeeSats;
  if (rateOut.integer <= 0n) return undefined;

  // pay_from = total_btc * (in/out) = total_sats * in / (out * 1e8).
  // Compute ceil(total_sats * in / out) as an 8-decimal fixed-point integer of
  // the from currency (i.e. units of 1e-8), then format — never binary floats.
  const payAt8Dp = ceilDiv(
    totalSats * rateIn.integer * rateOut.scale,
    rateIn.scale * rateOut.integer,
  );
  return formatDecimal(payAt8Dp, SATS_PER_BTC, 8);
}

/**
 * Maps XML from-side min/max into invoice-side msats using the pair's reference rate.
 * Minimum rounds up, maximum rounds down, so borderline invoices are never reported
 * as inside a range the provider would reject.
 */
export function invoiceLimitsFromFixedFloatRate(pair: FixedFloatRatePair): {
  readonly minimum_pay_amount: string;
  readonly maximum_pay_amount: string;
  readonly minimum_invoice_amount_msats?: number;
  readonly maximum_invoice_amount_msats?: number;
} {
  const minimumPayAmount = pair.minamount;
  const maximumPayAmount = pair.maxamount;
  const payUnitsPerSat = payUnitsPerInvoiceSat(pair);
  return {
    minimum_pay_amount: minimumPayAmount,
    maximum_pay_amount: maximumPayAmount,
    ...(payUnitsPerSat === undefined
      ? {}
      : {
          ...(payAmountToInvoiceMsats(minimumPayAmount, payUnitsPerSat, "ceil") === undefined
            ? {}
            : {
                minimum_invoice_amount_msats: payAmountToInvoiceMsats(
                  minimumPayAmount,
                  payUnitsPerSat,
                  "ceil",
                ),
              }),
          ...(payAmountToInvoiceMsats(maximumPayAmount, payUnitsPerSat, "floor") === undefined
            ? {}
            : {
                maximum_invoice_amount_msats: payAmountToInvoiceMsats(
                  maximumPayAmount,
                  payUnitsPerSat,
                  "floor",
                ),
              }),
        }),
  };
}

function payUnitsPerInvoiceSat(pair: FixedFloatRatePair): number | undefined {
  const rateIn = Number(pair.in);
  const outSats = btcAmountStringToSats(pair.out);
  if (!Number.isFinite(rateIn) || rateIn <= 0) return undefined;
  if (outSats === undefined || outSats <= 0) return undefined;
  return rateIn / outSats;
}

function payAmountToInvoiceMsats(
  payAmount: string,
  payUnitsPerSat: number,
  rounding: "ceil" | "floor",
): number | undefined {
  const value = Number(payAmount);
  if (!Number.isFinite(value) || value <= 0) return undefined;
  const invoiceSats = value / payUnitsPerSat;
  if (!Number.isFinite(invoiceSats) || invoiceSats <= 0) return undefined;
  const roundedSats = rounding === "ceil" ? Math.ceil(invoiceSats) : Math.floor(invoiceSats);
  const msats = roundedSats * 1000;
  return Number.isSafeInteger(msats) ? msats : undefined;
}

function btcAmountStringToSats(value: string): number | undefined {
  if (!DECIMAL_PATTERN.test(value)) return undefined;
  const [wholePart, fractionalPart = ""] = value.split(".");
  if (fractionalPart.length > 8) return undefined;
  const sats = BigInt(wholePart) * SATS_PER_BTC + BigInt(fractionalPart.padEnd(8, "0"));
  return sats <= BigInt(Number.MAX_SAFE_INTEGER) ? Number(sats) : undefined;
}

function parseToFeeBtcSats(tofee: string | undefined): bigint | undefined {
  if (tofee === undefined) return undefined;
  // Examples: "0.0004967000 BTC", "0.0005 BTCLN". Non-BTC fees are ignored —
  // we always pay out Lightning BTC, so only BTC network fees fold into pay-in.
  const match = tofee.trim().match(/^([0-9]+(?:\.[0-9]+)?)\s*([A-Za-z]+)?$/);
  if (match === null) return undefined;
  const amount = match[1];
  const unit = (match[2] ?? "BTC").toUpperCase();
  if (unit !== "BTC" && unit !== "BTCLN") return undefined;
  const parsed = parsePositiveDecimal(amount);
  if (parsed === undefined) return undefined;
  if (parsed.scale > SATS_PER_BTC) return undefined;
  return (parsed.integer * SATS_PER_BTC) / parsed.scale;
}

function parsePositiveDecimal(
  value: string,
): { integer: bigint; scale: bigint } | undefined {
  if (!DECIMAL_PATTERN.test(value)) return undefined;
  const [whole, fraction = ""] = value.split(".");
  const integer = BigInt(`${whole}${fraction}`);
  if (integer <= 0n) return undefined;
  return { integer, scale: 10n ** BigInt(fraction.length) };
}

function formatDecimal(integer: bigint, scale: bigint, maxFractionDigits: number): string {
  const whole = integer / scale;
  let fraction = integer % scale;
  // Truncate/pad to maxFractionDigits, then round up any discarded remainder.
  const targetScale = 10n ** BigInt(maxFractionDigits);
  if (scale > targetScale) {
    const divisor = scale / targetScale;
    const remainder = fraction % divisor;
    fraction = fraction / divisor;
    if (remainder > 0n) fraction += 1n;
    if (fraction >= targetScale) {
      return formatDecimal(whole * targetScale + fraction, targetScale, maxFractionDigits);
    }
    scale = targetScale;
  } else if (scale < targetScale) {
    fraction *= targetScale / scale;
    scale = targetScale;
  }
  const fractionText = fraction.toString().padStart(maxFractionDigits, "0").replace(/0+$/, "");
  return fractionText.length === 0 ? whole.toString() : `${whole.toString()}.${fractionText}`;
}

function ceilDiv(numerator: bigint, denominator: bigint): bigint {
  return (numerator + denominator - 1n) / denominator;
}

function stripCurrencySuffix(value: string): string {
  const match = value.trim().match(/^([0-9]+(?:\.[0-9]+)?)/);
  return match?.[1] ?? value.trim();
}

function matchTags(xml: string, tag: string): string[] {
  const pattern = new RegExp(`<${tag}\\b[^>]*>([\\s\\S]*?)</${tag}>`, "gi");
  const matches: string[] = [];
  for (const match of xml.matchAll(pattern)) {
    matches.push(match[1] ?? "");
  }
  return matches;
}

function readTagText(xml: string, tag: string): string | undefined {
  const pattern = new RegExp(`<${tag}\\b[^>]*>([\\s\\S]*?)</${tag}>`, "i");
  const match = xml.match(pattern);
  if (match === null) return undefined;
  const text = (match[1] ?? "").trim();
  return text.length === 0 ? undefined : text;
}

function readStoredPair(value: unknown): FixedFloatRatePair | undefined {
  if (value === null || typeof value !== "object") return undefined;
  const record = value as Record<string, unknown>;
  const from = typeof record.from === "string" ? record.from : undefined;
  const to = typeof record.to === "string" ? record.to : undefined;
  const inAmount = typeof record.in === "string" ? record.in : undefined;
  const outAmount = typeof record.out === "string" ? record.out : undefined;
  const amount = typeof record.amount === "string" ? record.amount : undefined;
  const minamount = typeof record.minamount === "string" ? record.minamount : undefined;
  const maxamount = typeof record.maxamount === "string" ? record.maxamount : undefined;
  if (
    from === undefined ||
    to === undefined ||
    inAmount === undefined ||
    outAmount === undefined ||
    amount === undefined ||
    minamount === undefined ||
    maxamount === undefined
  ) {
    return undefined;
  }
  return {
    from,
    to,
    in: inAmount,
    out: outAmount,
    amount,
    minamount,
    maxamount,
    ...(typeof record.tofee === "string" ? { tofee: record.tofee } : {}),
  };
}

File: packages/dotnet/BTCPayServer.Plugins.OpenReceive/Swaps/FixedFloatRates.cs
#nullable enable
using System;
using System.Collections.Generic;
using System.Numerics;
using System.Text.RegularExpressions;
using BTCPayServer.Plugins.OpenReceive.Generated;

namespace BTCPayServer.Plugins.OpenReceive.Swaps;

/// <summary>One pair from the FixedFloat public XML rates export.</summary>
/// <param name="In">Reference send amount in <paramref name="From"/> (defines the rate with <paramref name="Out"/>).</param>
/// <param name="Out">Reference receive amount in <paramref name="To"/>.</param>
/// <param name="Amount">Reserve availability hint for <paramref name="To"/> — not a max exchange amount.</param>
/// <param name="ToFee">Network fee on the <paramref name="To"/> side, excluded from <paramref name="Out"/> (e.g. "0.0005 BTC").</param>
public sealed record FixedFloatRatePair(
    string From,
    string To,
    string In,
    string Out,
    string Amount,
    string MinAmount,
    string MaxAmount,
    string? ToFee = null);

/// <summary>Pairs keyed as <c>FROM:TO</c> (uppercased), stamped with when they were fetched.</summary>
public sealed record FixedFloatRatesIndex(long FetchedAt, IReadOnlyDictionary<string, FixedFloatRatePair> Pairs);

/// <summary>Invoice-side limits derived from a pair's XML from-side min/max.</summary>
public sealed record FixedFloatInvoiceLimits(
    string MinimumPayAmount,
    string MaximumPayAmount,
    long? MinimumInvoiceAmountMsats,
    long? MaximumInvoiceAmountMsats);

/// <summary>
/// FixedFloat public XML rates export — the bulk feed for all pairs. No API key, no
/// weight budget. OpenReceive keeps only Lightning-payout pairs that match its small
/// pay-in asset list and derives indicative quotes / min-max locally. <c>/create</c>
/// remains authoritative.
/// </summary>
public static partial class FixedFloatRates
{
    /// <summary>Fraction digits the indicative pay-in amount is reported at.</summary>
    private const int PayAmountFractionDigits = 8;

    [GeneratedRegex(@"<item\b[^>]*>([\s\S]*?)</item>", RegexOptions.IgnoreCase)]
    private static partial Regex ItemPattern();

    [GeneratedRegex(@"^([0-9]+(?:\.[0-9]+)?)")]
    private static partial Regex LeadingDecimalPattern();

    [GeneratedRegex(@"^([0-9]+(?:\.[0-9]+)?)\s*([A-Za-z]+)?$")]
    private static partial Regex ToFeePattern();

    public static string PairKey(string from, string to) => $"{from.Trim().ToUpperInvariant()}:{to.Trim().ToUpperInvariant()}";

    public static string RatesXmlPath(string rateType = "fixed") => $"/rates/{rateType}.xml";

    /// <summary>Keep only pairs whose <c>to</c> side is a Lightning BTC payout code.</summary>
    public static IReadOnlyDictionary<string, FixedFloatRatePair> RetainLightningPayoutPairs(IReadOnlyDictionary<string, FixedFloatRatePair> pairs)
    {
        var retained = new Dictionary<string, FixedFloatRatePair>(StringComparer.Ordinal);
        foreach (var (key, pair) in pairs)
        {
            if (SwapAssets.IsLightningNetwork(pair.To)) retained[key] = pair;
        }
        return retained;
    }

    /// <summary>Keep only the from→Lightning keys that match resolved OpenReceive pay-in currencies.</summary>
    public static FixedFloatRatesIndex RetainPairsForKeys(FixedFloatRatesIndex index, IReadOnlySet<string> pairKeys)
    {
        var pairs = new Dictionary<string, FixedFloatRatePair>(StringComparer.Ordinal);
        if (pairKeys.Count == 0) return new FixedFloatRatesIndex(index.FetchedAt, pairs);
        foreach (var key in pairKeys)
        {
            if (index.Pairs.TryGetValue(key, out var pair)) pairs[key] = pair;
        }
        return new FixedFloatRatesIndex(index.FetchedAt, pairs);
    }

    public static IReadOnlyDictionary<string, FixedFloatRatePair> ParseRatesXml(string xml)
    {
        var pairs = new Dictionary<string, FixedFloatRatePair>(StringComparer.Ordinal);
        foreach (Match item in ItemPattern().Matches(xml))
        {
            var itemXml = item.Groups[1].Value;
            var from = ReadTagText(itemXml, "from");
            var to = ReadTagText(itemXml, "to");
            var inAmount = ReadTagText(itemXml, "in");
            var outAmount = ReadTagText(itemXml, "out");
            var amount = ReadTagText(itemXml, "amount");
            var minamount = ReadTagText(itemXml, "minamount");
            var maxamount = ReadTagText(itemXml, "maxamount");
            if (from is null || to is null || inAmount is null || outAmount is null ||
                amount is null || minamount is null || maxamount is null)
            {
                continue;
            }
            var tofee = ReadTagText(itemXml, "tofee");
            var pair = new FixedFloatRatePair(
                from.Trim(),
                to.Trim(),
                StripCurrencySuffix(inAmount),
                StripCurrencySuffix(outAmount),
                StripCurrencySuffix(amount),
                StripCurrencySuffix(minamount),
                StripCurrencySuffix(maxamount),
                tofee?.Trim());
            pairs[PairKey(pair.From, pair.To)] = pair;
        }
        return pairs;
    }

    /// <summary>
    /// Indicative pay-in amount for a Lightning payout of <paramref name="invoiceAmountMsats"/>:
    /// pay_from = (invoice_btc + tofee_btc) × (in / out), rounded UP at 8 decimals so the UI
    /// never understates what <c>/create</c> is likely to require. Null when the pair is unquotable.
    /// </summary>
    public static string? QuotePayAmountFromRate(FixedFloatRatePair pair, long invoiceAmountMsats)
    {
        if (invoiceAmountMsats <= 0 || invoiceAmountMsats > OpenReceiveTables.MaxAmountMsats) return null;
        var rateIn = ParsePositiveDecimal(pair.In);
        var rateOut = ParsePositiveDecimal(pair.Out);
        if (rateIn is null || rateOut is null) return null;

        var invoiceSats = DecimalMath.CeilDiv(new BigInteger(invoiceAmountMsats), DecimalMath.MsatsPerSat);
        var tofeeSats = ParseToFeeBtcSats(pair.ToFee) ?? BigInteger.Zero;
        var totalSats = invoiceSats + tofeeSats;

        // pay_from = total_sats * in / (out * 1e8), computed as an 8-decimal fixed-point
        // integer of the from currency — never binary floats.
        var payAt8Dp = DecimalMath.CeilDiv(
            totalSats * rateIn.Value.Units * DecimalMath.ScaleFactor(rateOut.Value.Scale),
            DecimalMath.ScaleFactor(rateIn.Value.Scale) * rateOut.Value.Units);
        return FormatPayAmount(payAt8Dp, PayAmountFractionDigits, PayAmountFractionDigits);
    }

    /// <summary>
    /// Maps XML from-side min/max into invoice-side msats using the pair's reference rate.
    /// Minimum rounds up, maximum rounds down, so borderline invoices are never reported
    /// as inside a range the provider would reject.
    /// </summary>
    public static FixedFloatInvoiceLimits InvoiceLimitsFromRate(FixedFloatRatePair pair) =>
        new(
            pair.MinAmount,
            pair.MaxAmount,
            PayAmountToInvoiceMsats(pair, pair.MinAmount, roundUp: true),
            PayAmountToInvoiceMsats(pair, pair.MaxAmount, roundUp: false));

    /// <summary>
    /// Compare two positive decimal strings: negative when left &lt; right, zero when equal,
    /// positive when left &gt; right. Null when either is not a positive decimal.
    /// </summary>
    public static int? CompareDecimalAmounts(string left, string right)
    {
        var a = ParsePositiveDecimal(left);
        var b = ParsePositiveDecimal(right);
        if (a is null || b is null) return null;
        var leftScaled = a.Value.Units * DecimalMath.ScaleFactor(b.Value.Scale);
        var rightScaled = b.Value.Units * DecimalMath.ScaleFactor(a.Value.Scale);
        return leftScaled.CompareTo(rightScaled) switch
        {
            < 0 => -1,
            > 0 => 1,
            _ => 0,
        };
    }

    /// <summary>Inverse of the direction=to quote (ignoring tofee): invoice_sats = pay_from × out × 1e8 / in.</summary>
    private static long? PayAmountToInvoiceMsats(FixedFloatRatePair pair, string payAmount, bool roundUp)
    {
        var pay = ParsePositiveDecimal(payAmount);
        var rateIn = ParsePositiveDecimal(pair.In);
        var rateOut = ParsePositiveDecimal(pair.Out);
        if (pay is null || rateIn is null || rateOut is null) return null;

        var numerator = pay.Value.Units * rateOut.Value.Units * DecimalMath.SatsPerBtc * DecimalMath.ScaleFactor(rateIn.Value.Scale);
        var denominator = DecimalMath.ScaleFactor(pay.Value.Scale) * DecimalMath.ScaleFactor(rateOut.Value.Scale) * rateIn.Value.Units;
        if (denominator.Sign <= 0) return null;
        var invoiceSats = roundUp ? DecimalMath.CeilDiv(numerator, denominator) : numerator / denominator;
        if (invoiceSats.Sign <= 0) return null;
        var msats = invoiceSats * DecimalMath.MsatsPerSat;
        if (msats > new BigInteger(OpenReceiveTables.MaxAmountMsats)) return null;
        return (long)msats;
    }

    private static BigInteger? ParseToFeeBtcSats(string? tofee)
    {
        if (tofee is null) return null;
        // Examples: "0.0004967000 BTC", "0.0005 BTCLN". Non-BTC fees are ignored —
        // we always pay out Lightning BTC, so only BTC network fees fold into pay-in.
        var match = ToFeePattern().Match(tofee.Trim());
        if (!match.Success) return null;
        var amount = match.Groups[1].Value;
        var unit = (match.Groups[2].Success ? match.Groups[2].Value : "BTC").ToUpperInvariant();
        if (unit != "BTC" && unit != "BTCLN") return null;
        var parsed = ParsePositiveDecimal(amount);
        if (parsed is null) return null;
        // Fees carrying more than 8 decimals are reduced to whole sats with ceil rounding.
        return DecimalMath.CeilDiv(parsed.Value.Units * DecimalMath.SatsPerBtc, DecimalMath.ScaleFactor(parsed.Value.Scale));
    }

    /// <summary>Provider-data decimal parse: an unusable value is "cannot quote" (null), never a throw.</summary>
    private static DecimalUnits? ParsePositiveDecimal(string value)
    {
        DecimalUnits parsed;
        try
        {
            parsed = DecimalMath.ParseDecimal(value);
        }
        catch (FormatException)
        {
            return null;
        }
        return parsed.Units.Sign > 0 ? parsed : null;
    }

    /// <summary>
    /// Render fixed-point units at scale with at most maxFractionDigits, rounding any
    /// discarded remainder up, and trimming trailing zeros.
    /// </summary>
    private static string FormatPayAmount(BigInteger units, int scale, int maxFractionDigits)
    {
        var target = DecimalMath.ScaleFactor(maxFractionDigits);
        var current = DecimalMath.ScaleFactor(scale);
        var rescaled = scale <= maxFractionDigits ? units * (target / current) : DecimalMath.CeilDiv(units * target, current);
        var text = DecimalMath.FormatDecimal(rescaled, maxFractionDigits);
        return text.Contains('.') ? text.TrimEnd('0').TrimEnd('.') : text;
    }

    private static string StripCurrencySuffix(string value)
    {
        var trimmed = value.Trim();
        var match = LeadingDecimalPattern().Match(trimmed);
        return match.Success ? match.Groups[1].Value : trimmed;
    }

    private static string? ReadTagText(string xml, string tag)
    {
        var match = Regex.Match(xml, $@"<{tag}\b[^>]*>([\s\S]*?)</{tag}>", RegexOptions.IgnoreCase);
        if (!match.Success) return null;
        var text = match.Groups[1].Value.Trim();
        return text.Length == 0 ? null : text;
    }
}
